File: web/src/source/mockSignal.ts
/**
 * Synthetic biosignal generator (mock mode, §8). Produces realistic ECG (QRS
 * morphology), PPG (pulse wave whose FOOT is phase-lagged from the R-peak by an exact
 * PAT), and BioZ (stable Z₀ + small pulsatile ΔZ, with injectable motion bursts).
 *
 * Ground truth is exact and deterministic: R-peaks occur at k·RR and PPG feet at
 * k·RR + PAT, so the DSP can be unit-tested against known HR / PAT / SpO₂.
 */
import { US_PER_S } from '../dsp/constants';

export interface MockParams {
  hrBpm: number;
  patMs: number;
  ecgRateHz: number;
  ppgRateHz: number;
  biozRateHz: number;
  ecgAmplitude: number;
  ecgNoise: number;
  ppgGreenDc: number;
  ppgGreenAc: number;
  /** Target SpO₂ used to set the red/IR AC ratio (ground truth for the SpO₂ test). */
  spo2Target: number;
  z0Baseline: number;
  z0Noise: number;
  dzAmplitude: number;
  dzNoise: number;
  seed: number;
  /** Motion bursts inject high-variance ΔZ (drives the motion/gating logic). */
  motionBursts: { startUs: number; endUs: number; severity: number }[];
}

export const DEFAULT_MOCK_PARAMS: MockParams = {
  hrBpm: 70,
  patMs: 220,
  ecgRateHz: 256,
  ppgRateHz: 100,
  biozRateHz: 64,
  ecgAmplitude: 1200,
  ecgNoise: 8,
  ppgGreenDc: 120000,
  ppgGreenAc: 9000,
  spo2Target: 98,
  z0Baseline: 300000,
  z0Noise: 1500,
  dzAmplitude: 400,
  dzNoise: 60,
  seed: 1,
  motionBursts: [],
};

/** Deterministic value-noise in [-1, 1] from a real-valued coordinate. */
function valNoise(x: number): number {
  const s = Math.sin(x * 12.9898) * 43758.5453;
  return 2 * (s - Math.floor(s)) - 1;
}

export class MockSignal {
  readonly p: MockParams;
  private readonly rrUs: number;

  constructor(params: Partial<MockParams> = {}) {
    this.p = { ...DEFAULT_MOCK_PARAMS, ...params };
    this.rrUs = (60 / this.p.hrBpm) * US_PER_S;
  }

  get rrMicros(): number {
    return this.rrUs;
  }

  /** R-peak timestamps (µs) within [t0, t1) — the exact PAT/HR ground truth. */
  rPeakTimes(t0: number, t1: number): number[] {
    const out: number[] = [];
    const kStart = Math.ceil(t0 / this.rrUs);
    const kEnd = Math.floor((t1 - 1) / this.rrUs);
    for (let k = kStart; k <= kEnd; k++) out.push(k * this.rrUs);
    return out;
  }

  /** PPG foot timestamps (µs) within [t0, t1). */
  footTimes(t0: number, t1: number): number[] {
    const patUs = this.p.patMs * 1000;
    return this.rPeakTimes(t0 - patUs, t1 - patUs).map((r) => r + patUs);
  }

  ecgAt(tUs: number): number {
    const k0 = Math.floor(tUs / this.rrUs);
    let v = 0;
    for (let k = k0 - 1; k <= k0 + 1; k++) {
      v += qrs((tUs - k * this.rrUs) / US_PER_S);
    }
    return v * this.p.ecgAmplitude + this.p.ecgNoise * valNoise(this.p.seed + tUs * 1e-3);
  }

  greenAt(tUs: number): number {
    return this.p.ppgGreenDc + this.p.ppgGreenAc * this.pulseSum(tUs) +
      this.p.ppgGreenAc * 0.01 * valNoise(this.p.seed + 7 + tUs * 1e-3);
  }

  redAt(tUs: number): number {
    // R = (AC_red/DC_red)/(AC_ir/DC_ir); with equal DCs, AC_red/AC_ir = R = (110-SpO2)/25.
    const r = (110 - this.p.spo2Target) / 25;
    const ac = this.p.ppgGreenAc * r;
    return this.p.ppgGreenDc + ac * this.pulseSum(tUs) +
      ac * 0.01 * valNoise(this.p.seed + 11 + tUs * 1e-3);
  }

  irAt(tUs: number): number {
    const ac = this.p.ppgGreenAc;
    return this.p.ppgGreenDc + ac * this.pulseSum(tUs) +
      ac * 0.01 * valNoise(this.p.seed + 13 + tUs * 1e-3);
  }

  z0At(tUs: number): number {
    const drift = 4000 * Math.sin((2 * Math.PI * tUs) / (20 * US_PER_S));
    return this.p.z0Baseline + drift + this.p.z0Noise * valNoise(this.p.seed + 17 + tUs * 1e-3);
  }

  dzAt(tUs: number): number {
    const patUs = this.p.patMs * 1000;
    const pulsatile = this.p.dzAmplitude * this.pulseSum(tUs - patUs);
    let motion = 0;
    for (const b of this.p.motionBursts) {
      if (tUs >= b.startUs && tUs < b.endUs) {
        motion += b.severity * 3000 * valNoise(this.p.seed + 23 + tUs * 1e-3);
      }
    }
    return pulsatile + this.p.dzNoise * valNoise(this.p.seed + 19 + tUs * 1e-3) + motion;
  }

  /** Sum of the pulse waveform over nearby beats; foot of beat k is at k·RR + PAT. */
  private pulseSum(tUs: number): number {
    const patUs = this.p.patMs * 1000;
    const k0 = Math.floor((tUs - patUs) / this.rrUs);
    let v = 0;
    for (let k = k0 - 1; k <= k0 + 1; k++) {
      const foot = k * this.rrUs + patUs;
      v += pulse((tUs - foot) / US_PER_S);
    }
    return v;
  }
}

/** QRS-ish morphology: dominant R at τ=0, flanking Q/S, broad T, small P. τ in seconds. */
function qrs(tau: number): number {
  const g = (a: number, c: number, w: number) => a * Math.exp(-(((tau - c) / w) ** 2));
  return (
    g(1.0, 0.0, 0.012) + // R
    g(-0.15, -0.022, 0.012) + // Q
    g(-0.2, 0.022, 0.014) + // S
    g(0.22, 0.3, 0.06) + // T
    g(0.08, -0.18, 0.04) // P
  );
}

/** Alpha-function pulse with onset (foot) exactly at τ=0, plus a small dicrotic bump. */
function pulse(tau: number): number {
  if (tau < 0) return 0;
  const t0 = 0.12; // systolic peak time (s)
  const alpha = (t: number, k: number) => (t < 0 ? 0 : (t / k) * Math.exp(1 - t / k));
  return alpha(tau, t0) + 0.25 * alpha(tau - 0.32, 0.1);
}

File: android/app/src/main/java/com/monivitals/source/MockSignal.kt
//
//  MockSignal.kt
//  MoniVitals (Android)
//
//  Synthetic biosignal generator (mock mode). Produces realistic ECG (QRS morphology),
//  PPG (pulse wave whose FOOT is phase-lagged from the R-peak by an exact PAT), and BioZ
//  (stable Z0 + small pulsatile ΔZ, with injectable motion bursts).
//
//  Ground truth is exact and deterministic: R-peaks occur at k·RR and PPG feet at
//  k·RR + PAT, so the DSP can be unit-tested against known HR / PAT / SpO2.
//
//  Ported from ios/MoniVitals/Source/MockSignal.swift (== web/src/source/mockSignal.ts).
//

package com.monivitals.source

import com.monivitals.dsp.DSPConstants
import kotlin.math.PI
import kotlin.math.ceil
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.pow
import kotlin.math.sin

/** A high-variance ΔZ injection window driving the motion/gating logic. */
data class MotionBurst(
    val startUs: Double,
    val endUs: Double,
    val severity: Double,
)

/** Tunable parameters of the synthetic signal. */
data class MockParams(
    var hrBpm: Double,
    var patMs: Double,
    var ecgRateHz: Int,
    var ppgRateHz: Int,
    var biozRateHz: Int,
    var ecgAmplitude: Double,
    var ecgNoise: Double,
    var ppgGreenDc: Double,
    var ppgGreenAc: Double,
    /** Target SpO2 used to set the red/IR AC ratio (ground truth for the SpO2 test). */
    var spo2Target: Double,
    var z0Baseline: Double,
    var z0Noise: Double,
    var dzAmplitude: Double,
    var dzNoise: Double,
    var seed: Double,
    /** Motion bursts inject high-variance ΔZ (drives the motion/gating logic). */
    var motionBursts: MutableList<MotionBurst>,
) {
    companion object {
        val default: MockParams
            get() = MockParams(
                hrBpm = 70.0,
                patMs = 220.0,
                ecgRateHz = 256,
                ppgRateHz = 100,
                biozRateHz = 64,
                ecgAmplitude = 1200.0,
                ecgNoise = 8.0,
                ppgGreenDc = 120_000.0,
                ppgGreenAc = 9000.0,
                spo2Target = 98.0,
                z0Baseline = 300_000.0,
                z0Noise = 1500.0,
                dzAmplitude = 400.0,
                dzNoise = 60.0,
                seed = 1.0,
                motionBursts = mutableListOf(),
            )
    }
}

/** Deterministic value-noise in [-1, 1] from a real-valued coordinate. */
private fun valNoise(x: Double): Double {
    val s = sin(x * 12.9898) * 43758.5453
    return 2 * (s - floor(s)) - 1
}

class MockSignal(var p: MockParams = MockParams.default) {

    /** R–R interval (µs); recomputed from the current HR so it tracks live changes. */
    private val rrUs: Double
        get() = (60 / p.hrBpm) * DSPConstants.usPerS

    val rrMicros: Double get() = rrUs

    /** R-peak timestamps (µs) within `[t0, t1)` — the exact PAT/HR ground truth. */
    fun rPeakTimes(t0: Double, t1: Double): List<Double> {
        val out = mutableListOf<Double>()
        val kStart = ceil(t0 / rrUs).toInt()
        val kEnd = floor((t1 - 1) / rrUs).toInt()
        if (kEnd >= kStart) {
            for (k in kStart..kEnd) out.add(k * rrUs)
        }
        return out
    }

    /** PPG foot timestamps (µs) within `[t0, t1)`. */
    fun footTimes(t0: Double, t1: Double): List<Double> {
        val patUs = p.patMs * 1000
        return rPeakTimes(t0 - patUs, t1 - patUs).map { it + patUs }
    }

    fun ecgAt(tUs: Double): Double {
        val k0 = floor(tUs / rrUs).toInt()
        var v = 0.0
        for (k in (k0 - 1)..(k0 + 1)) {
            v += qrs((tUs - k * rrUs) / DSPConstants.usPerS)
        }
        return v * p.ecgAmplitude + p.ecgNoise * valNoise(p.seed + tUs * 1e-3)
    }

    fun greenAt(tUs: Double): Double {
        return p.ppgGreenDc + p.ppgGreenAc * pulseSum(tUs) +
            p.ppgGreenAc * 0.01 * valNoise(p.seed + 7 + tUs * 1e-3)
    }

    fun redAt(tUs: Double): Double {
        // R = (AC_red/DC_red)/(AC_ir/DC_ir); with equal DCs, AC_red/AC_ir = R = (110-SpO2)/25.
        val r = (110 - p.spo2Target) / 25
        val ac = p.ppgGreenAc * r
        return p.ppgGreenDc + ac * pulseSum(tUs) +
            ac * 0.01 * valNoise(p.seed + 11 + tUs * 1e-3)
    }

    fun irAt(tUs: Double): Double {
        val ac = p.ppgGreenAc
        return p.ppgGreenDc + ac * pulseSum(tUs) +
            ac * 0.01 * valNoise(p.seed + 13 + tUs * 1e-3)
    }

    fun z0At(tUs: Double): Double {
        val drift = 4000 * sin((2 * PI * tUs) / (20 * DSPConstants.usPerS))
        return p.z0Baseline + drift + p.z0Noise * valNoise(p.seed + 17 + tUs * 1e-3)
    }

    fun dzAt(tUs: Double): Double {
        val patUs = p.patMs * 1000
        val pulsatile = p.dzAmplitude * pulseSum(tUs - patUs)
        var motion = 0.0
        for (b in p.motionBursts) {
            if (tUs >= b.startUs && tUs < b.endUs) {
                motion += b.severity * 3000 * valNoise(p.seed + 23 + tUs * 1e-3)
            }
        }
        return pulsatile + p.dzNoise * valNoise(p.seed + 19 + tUs * 1e-3) + motion
    }

    /** Sum of the pulse waveform over nearby beats; foot of beat k is at k·RR + PAT. */
    private fun pulseSum(tUs: Double): Double {
        val patUs = p.patMs * 1000
        val k0 = floor((tUs - patUs) / rrUs).toInt()
        var v = 0.0
        for (k in (k0 - 1)..(k0 + 1)) {
            val foot = k * rrUs + patUs
            v += pulse((tUs - foot) / DSPConstants.usPerS)
        }
        return v
    }
}

/** QRS-ish morphology: dominant R at τ=0, flanking Q/S, broad T, small P. τ in seconds. */
private fun qrs(tau: Double): Double {
    fun g(a: Double, c: Double, w: Double): Double = a * exp(-((tau - c) / w).pow(2))
    return g(1.0, 0.0, 0.012) +      // R
        g(-0.15, -0.022, 0.012) +    // Q
        g(-0.2, 0.022, 0.014) +      // S
        g(0.22, 0.3, 0.06) +         // T
        g(0.08, -0.18, 0.04)         // P
}

/** Alpha-function pulse with onset (foot) exactly at τ=0, plus a small dicrotic bump. */
private fun pulse(tau: Double): Double {
    if (tau < 0) return 0.0
    val t0 = 0.12 // systolic peak time (s)
    fun alpha(t: Double, k: Double): Double = if (t < 0) 0.0 else (t / k) * exp(1 - t / k)
    return alpha(tau, t0) + 0.25 * alpha(tau - 0.32, 0.1)
}
